labgrid/driver/shelldriver.py

```diff
@@ -236,11 +236,24 @@ def _check_prompt(self):
             self._status = 0
             raise
 
-    def _inject_run(self):
-        self.console.sendline(
-            '''run() { echo -n "$MARKER"; sh -c "$@"; echo "$MARKER $?"; }'''
-        )
-        self.console.expect(self.prompt)
+    def _inject_run(self, inject_timeout=10.0):
+        """Tries to inject run() command wrapper into shell on the console."""
+        timeout = Timeout(inject_timeout)
+
+        while True:
+            self.console.sendline(
+                '''run() { echo -n "$MARKER"; sh -c "$@"; echo "$MARKER $?"; }'''
+            )
+            self.console.expect(self.prompt)
+            try:
+                data = self._raw_run("type run")
+                if data[2] and "function" in data[2]:
+                    return
+            except TIMEOUT:
+                pass
+
+            if timeout.expired:
+                raise TIMEOUT(f"Timeout of {inject_timeout} seconds exceeded during waiting for run injection")  # pylint: disable=line-too-long
```

Comment:
If your serial console does not work reliably (even after specifying a certain settle time), I don't think retrying is the correct approach. Even if this works after the nth try, too many other things will fail in unexpected ways later. Better make your serial console work reliably or prefer the SSHDriver in these situations.

Reply:
> If your serial console does not work reliably (even after specifying a certain settle time)

Ok, fair enough, but my goal is to make it generic and reusable across different use cases, instead of fiddling with variable settle time, fighting crazy applications logging into `/dev/console` directly etc.

> If your serial console does not work reliably (even after specifying a certain settle time), I don't think retrying is the correct approach. 

Well, this approach works reliably for over than 2 years, with a mix of dozen different devices, bootloaders, kernels/drivers and operating systems. So the idea here is to get as fast (and reliably) as possible into the SSH and profit.   

> SSHDriver in these situations.

Indeed, its the only reliable solution if there is no dedicated serial console available.